include/ekuvector/ekuvector.hpp

```diff
@@ -0,0 +1,1122 @@
+  template <class... Args> void emplace_back(Args &&... args);
+
+  /** @brief Removes the last element of the container.
+   *
+   * Calling pop_back on an empty container is undefined. Iterators and
+   * references to the last element, as well as the end() iterator, are
+   * invalidated. */
+  void pop_back();
+
+  /** @brief Resizes the container to contain count elements.
+   *
+   * If the current size is greater than count, the container is reduced to its
+   * first count elements. If the current size is less than count, additional
+   * elements are appended and initialized with copies of value. If the current
+   * size is less than count, additional default-inserted elements are appended
+   */
+  void resize(size_type count);
+
+  /** @brief Resizes the container to contain count elements.
+   *
+   * If the current size is greater than count, the container is reduced to its
+   * first count elements. If the current size is less than count, additional
+   * elements are appended and initialized with copies of value. If the current
+   * size is less than count, additional copies of value are appended */
+  void resize(size_type count, const value_type &value);
+
+  /** @brief Exchanges the contents of the container with those of other.
+   *
+   * Does not invoke any move, copy, or swap operations on individual elements.
+   * All iterators and references remain valid. The past-the-end iterator is
+   * invalidated. */
+  void swap(ekuvector &other);
+
+private:
+  Allocator allocator_;
+  size_t capacity_;
+  size_t size_;
+  pointer data_;
+
+  void preallocate_capacity(size_type new_cap);
+};
+
+template <class Type, class Allocator>
+ekuvector<Type, Allocator>::ekuvector() : ekuvector(Allocator()) {}
+
+template <class Type, class Allocator>
+ekuvector<Type, Allocator>::ekuvector(const Allocator &alloc)
+    : allocator_{alloc}, capacity_{0}, size_{0}, data_{nullptr} {}
+
+template <class Type, class Allocator>
+ekuvector<Type, Allocator>::ekuvector(size_type count, const Type &value,
+                                      const Allocator &alloc)
+    : ekuvector(alloc) {
+  if (count) {
+    /* preallocate memory space for count elements */
+    preallocate_capacity(count);
+    /* call a copy constructor on each */
+    while (size_ < count) {
+      emplace_back(value);
+    }
+  }
+}
+
+template <class Type, class Allocator>
+ekuvector<Type, Allocator>::ekuvector(size_type count) : ekuvector() {
+  if (count) {
+    /* preallocate memory space for count elements */
+    preallocate_capacity(count);
+    /* call a default constructor on each new item */
+    while (size_ < count) {
+      emplace_back();
+    }
+  }
+}
+
+template <class Type, class Allocator>
+template <class InputIt>
+ekuvector<Type, Allocator>::ekuvector(
+    InputIt first,
+    typename std::enable_if<is_iterator<InputIt>::value, InputIt>::type last,
+    const Allocator &alloc)
+    : ekuvector(alloc) {
+  for (auto it = first; it != last; ++it) {
```

Comment:
Would std::distance() improve performance here?

Reply:
Btw, it's a nit.

Reply:
Do you mean using it to determine the stop condition instead of `it != last`? No, most likely it would not improve it, since `std::distance()` can be an expensive operation depending on what the type of the input iterators is.

For non-random-access iterators, `std::distance` is `O(d)`, where `d` is the actual distance, between the arguments to the function. This is so because in that case `std::distance` will use the auto-increment operator repeatedly to count the number of steps between its arguments. 

In that case the loop would become `O(n^2)` instead of `O(n)` as it is now.

Reply:
I did not have that in mind when asked about performance. My bad not being explicit before. I was thinking of `reserving()` memory first to later `emplace()` elements.
You can later iterate `first` and `last`.

Reply:
Now I see. In that case, yes, you're right, preallocating the size would result in better performance for large vectors.